### PR TITLE
Add workflows for building+publishing images & creating releases

### LINKED ISSUE #17: Create GitHub workflow for building images
After the selkies-gstreamer integration is done (#14), build a workflow to build all of the images and push them to GHCR with versions.

Follow the same build philosophy as the selkies-gstreamer repo:
https://github.com/selkies-project/selkies-gstreamer/tree/master/.github/workflows

Preserve the cloudbuild image building to allow folks to build and self-host images.

Strategy:

- master and dev branches are built when pushed to, images are tagged with the branch name respectively.
- Pushing a tag builds the images, creates a release and tags the images with latest

### PR DESCRIPTION
**Pull request by kili-ilo**
This completes most of #17. #24 will complete the last bit.

### CODE REVIEW

**Comment by kili-ilo** [comment]
I'm not sure how to test these, as the Actions console only supports running workflows that are already on the default branch.

**Comment by kili-ilo** [comment]
We could simply merge this PR and debug from there 🤷🏻

**Comment by kili-ilo** [comment]
After cursory research, it sounds like the prevailing wisdom is to optimistically merge all manually-triggered workflows, then later debug their behavior semi-manually while writing a second PR for the triggers.

**Comment by kili-ilo** [comment]
This PR is now safe to merge - after merging, I can test the new workflows and PR the on-push triggers for the image workflows.

File: .github/workflows/build_and_publish_all_images.yaml [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+jobs:
+
+  build_and_publish_app_proxy_image:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: ./.github/actions/build_and_publish_image
+      with:
+        environment: ${{ inputs.environment }}
+        image_source_directory: images/app-proxy
```
Comment (kili-ilo) [inline .github/workflows/build_and_publish_all_images.yaml]: This image didn't seem to be built by cloudbuild - instead, the build steps appear in https://github.com/selkies-project/selkies-vdi/blob/master/images/app-proxy/README.md . I assumed we should build & publish it anyways.

File: .github/workflows/build_and_publish_all_images.yaml [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    - uses: actions/checkout@v2
+    - uses: ./.github/actions/build_and_publish_image
+      with:
+        dockerfile: Dockerfile.bionic
+        environment: ${{ inputs.environment }}
+        image_source_directory: images/app-streaming
+        # TODO: Should we add "bionic" to the name?
+        image_name: webrtc-gpu-streaming-app-streaming
+        # TODO: Should we have a version history (with "latest")?
+        image_version_1: bionic
```
Comment (kili-ilo) [inline .github/workflows/build_and_publish_all_images.yaml]: Do we want a `:latest` for the bionic image? Should the `bionic` descriptor move before the tag's `:`?

File: .github/actions/build_and_publish_image/action.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      if [[ -n "${{ inputs.build_args }}" ]]; then
+        IFS=";" read -ra args <<< "${{ inputs.build_args }}"
+        for arg in ${args[@]}; do BUILD_ARGS+="--build-arg $arg "; done
+      fi
+
+      export IMAGE_TAG_URL_PREFIX=ghcr.io/selkies-project/selkies-vdi/
+      if [[ "${{ inputs.environment }}" == "master" ]]; then
+        export IMAGE_TAG_ENVIRONMENT_SUFFIX=""
+      else
+        export IMAGE_TAG_ENVIRONMENT_SUFFIX="-dev"
```
Comment (kili-ilo) [inline .github/actions/build_and_publish_image/action.yaml]: The only distinction between `dev` and `master` images is that `dev` image tags have a `-dev` suffix (ex: `ghcr.io/selkies-project/selkies-vdi/webrtc-gpu-streaming-app-streaming:latest-dev`

Comment (kili-ilo) [reply .github/actions/build_and_publish_image/action.yaml]: This is a mistake - the release version is already the branch name.

**Comment by kili-ilo** [comment]
~Typos! Replace `itopia-inc` with `selkies-project`, and replace `main` with `master`.~ Done.

File: .github/workflows/build_and_publish_all_images.yaml (hunk 1 above)
Comment (kili-ilo) [reply .github/workflows/build_and_publish_all_images.yaml]: Confirmed.

File: .github/workflows/build_and_publish_all_images.yaml [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+  build_and_publish_app_proxy_image:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: ./.github/actions/build_and_publish_image
+      with:
+        environment: ${{ inputs.environment }}
+        image_source_directory: images/app-proxy
+        image_name: webrtc-gpu-streaming-app-proxy
```
Comment (kili-ilo) [inline .github/workflows/build_and_publish_all_images.yaml]: Replace all instances of `webrtc-gpu-streaming-` in image tag basenames with ``. (Remove legacy prefixes).

File: .github/workflows/build_and_publish_all_images.yaml [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    - uses: actions/checkout@v2
+    - uses: ./.github/actions/build_and_publish_image
+      with:
+        environment: ${{ inputs.environment }}
+        image_source_directory: images/desktop
+        image_name: webrtc-gpu-streaming-desktop
+        image_version_1: ${{ inputs.image_version_1 }}
+        image_version_2: ${{ inputs.image_version_2 }}
+
+  build_and_publish_proton_app_image:
```
Comment (kili-ilo) [inline .github/workflows/build_and_publish_all_images.yaml]: Remove all proton-related jobs (it's unused and likely broken).

File: .github/workflows/build_and_publish_all_images.yaml [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    - uses: actions/checkout@v2
+    - uses: ./.github/actions/build_and_publish_image
+      with:
+        environment: ${{ inputs.environment }}
+        image_source_directory: images/xserver
+        image_name: webrtc-gpu-streaming-xserver
+        image_version_1: ${{ inputs.image_version_1 }}
+        image_version_2: ${{ inputs.image_version_2 }}
+
+  build_and_publish_webrtc_image:
```
Comment (kili-ilo) [inline .github/workflows/build_and_publish_all_images.yaml]: Alphabetize this job.

File: .github/workflows/build_and_publish_all_images.yaml (hunk 2 above)
Comment (kili-ilo) [reply .github/workflows/build_and_publish_all_images.yaml]: No. `:latest-bionic`, `:latest-buster`, etc.

**Comment by kili-ilo** [comment]
@danisla I've manually tested these workflows via the Actions web console.

File: .github/workflows/on_release_tag.yaml [hunk 7]
```diff
@@ -1,10 +1,27 @@
...
+  create_release:
+    needs: build_and_publish_all_images
     runs-on: ubuntu-latest
     steps:
-    - run: echo 'stub'
+    - uses: actions/create-release@v1
+      env:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        draft: false
```
Comment (danisla) [inline .github/workflows/on_release_tag.yaml]: What are your thoughts on setting `draft: true` to force us to create meaningful release notes?

File: .github/workflows/build_and_publish_changed_images.yaml [hunk 8]
```diff
@@ -1,10 +1,357 @@
...
+        github_token: $github_token
+        github_username: $github_username
+        image_name: app-streaming
+        image_source_directory: images/app-streaming
+        image_version_1: ${{ inputs.image_version_1 }}
+        image_version_suffix: -bionic
+
+  app_streaming_buster:
+    needs: check_image_source_changes
+    if: needs.check_image_source_changes.outputs.app_streaming_changed_buster == 'true'
```
Comment (danisla) [inline .github/workflows/build_and_publish_changed_images.yaml]: typo, change `app_streaming_changed_buster` to `app_streaming_buster_changed`

**Review by danisla — CHANGES_REQUESTED** [review]

File: .github/workflows/on_release_tag.yaml (hunk 7 above)
Comment (kili-ilo) [reply .github/workflows/on_release_tag.yaml]: Sounds good

File: .github/workflows/build_and_publish_changed_images.yaml (hunk 8 above)
Comment (kili-ilo) [reply .github/workflows/build_and_publish_changed_images.yaml]: Whoops, good catch

File: .github/workflows/on_release_tag.yaml (hunk 7 above)
Comment (kili-ilo) [reply .github/workflows/on_release_tag.yaml]: Done

File: .github/workflows/build_and_publish_changed_images.yaml (hunk 8 above)
Comment (kili-ilo) [reply .github/workflows/build_and_publish_changed_images.yaml]: Done

**Comment by kili-ilo** [comment]
@danisla I implemented those changes.

**Pull request merged**